### PR TITLE
Provence

### PR DESCRIPTION
**Pull request by nadiinchi**
* add Provence context_processor and configs
* add baseline context pruners context_processors and configs
* add context compression calculation & saving in a separate metric file in an exp folder
* little refactor of `modules/rag.py` to better accommodate context processing
* add possibility  to provide oracle docs in the query dataset
* add RGB dataset (with oracle docs in query dataset)

### CODE REVIEW

**Comment by DRRV** [comment]
just be to usre:  here Provence  does not do reranking right?

File: models/context_processors/recomp.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
+from torch.utils.data import Dataset, DataLoader
+from transformers import AutoTokenizer, AutoModel
+from transformers import T5ForConditionalGeneration
+from models.context_processors.context_processor import ContextProcessor
+import torch.nn.functional as F
+import torch
+from typing import List
+from llmlingua import PromptCompressor
```
Comment (sclincha) [inline models/context_processors/recomp.py:8]: Is this required? LLMlingua?

**Comment by nadiinchi** [comment]
> just be to usre: here Provence does not do reranking right?

it does. If `reorder=True` in provence config, provence reorders documents in the context processing list. But it is incorporated into context processing, so you would set `retriever=<smth> context_processor=provence/provence_reranking_0.1` and skip reranker

File: models/context_processors/recomp.py (hunk 1 above)
Comment (nadiinchi) [reply models/context_processors/recomp.py]: > Is this required? LLMlingua?

good point, thanks, fixed the imports!

also added a script `scripts/run_exps.sh` which  shows how to run exps, e.g. how to run Provence with reranking

File: models/context_processors/dslr_ce.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import torch.nn.functional as F
+import torch
+from typing import List
+from tqdm import tqdm
+import nltk
+import string
+import numpy as np
+
+
+class DSLR_CE(ContextProcessor):
```
Comment (maxime-louis) [inline models/context_processors/dslr_ce.py:13]: Maybe add a reference ? I don't know what that is

Comment (maxime-louis) [reply models/context_processors/dslr_ce.py:13]: or at least a docstring for the class !

File: models/context_processors/dslr_ce.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    def __init__(self, model_name="BAAI/bge-reranker-v2-m3", batch_size=32, max_len=512, threshold=None, always_select_title=True):
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        self.tokenizer = AutoTokenizer.from_pretrained(model_name)
+        self.model = AutoModelForSequenceClassification.from_pretrained(model_name, low_cpu_mem_usage=True, torch_dtype=torch.float16).to(self.device)
+
```
Comment (maxime-louis) [inline models/context_processors/dslr_ce.py:18]: Two consecutive line returns: only one is allowed within functions

File: models/context_processors/dslr_ce.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        scores = {qi:{ci:{} for ci in range(len(contexts[qi]))} for qi in range(len(queries))}
+        for i, (qi, ci, si) in enumerate(zip(qis, cis, sis)):
+            scores[qi][ci][si] = probs[i]
+        selected_contexts = []
+        for qi in range(len(queries)):
+            selected_contexts_ = []
+            for ci in range(len(contexts[qi])):
+                scores_ = scores[qi][ci]
+                
+                selected_idxs = [t for t in range(len(scores_)) if scores_[t+1]>self.threshold]
```
Comment (maxime-louis) [inline models/context_processors/dslr_ce.py:74]: the 't+1' will make an indexing error no ?

File: models/context_processors/dslr_ce.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        with torch.no_grad():
+            for batch_start in tqdm(range(0, len(context_sents_withtitle), self.batch_size), desc='Encoding contexts for pruning...'):
+                input_ids = toks['input_ids'][batch_start: min(batch_start+self.batch_size, len(context_sents_withtitle))].to(self.device)
+                attention_mask = toks['attention_mask'][batch_start: min(batch_start+self.batch_size, len(context_sents_withtitle))].to(self.device)
+                rank_score = self.model(input_ids=input_ids,
+                                attention_mask=attention_mask,
+                                ).logits
+                # Use sigmoid since it's BCEWithLogitsLoss
+                prob = torch.sigmoid(rank_score)
+                probs+=(prob)
```
Comment (maxime-louis) [inline models/context_processors/dslr_ce.py:63]: I would use 'prob.item()' to make it a float and not a torch tensor

File: models/context_processors/llmlingua2.py [hunk 6]
```diff
@@ -3,21 +3,23 @@
 from llmlingua import PromptCompressor
 from tqdm import tqdm
 
+
 class LLMLingua2(ContextProcessor):
```
Comment (maxime-louis) [inline models/context_processors/llmlingua2.py:7]: Arxiv link ?

File: models/context_processors/provence.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+class ProvenceCompressor(ContextProcessor):
+    def __init__(
+        self,
+        model_name,
+        name = "provence",
+        **kwargs
+    ):
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        self.model = AutoModel.from_pretrained(model_name, trust_remote_code=True).to(self.device)
+        self.kwargs = kwargs
```
Comment (maxime-louis) [inline models/context_processors/provence.py:16]: Passing kwargs like this is a bit strange, is it really convenient ?

File: models/context_processors/recomp.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+from transformers import T5ForConditionalGeneration
+from models.context_processors.context_processor import ContextProcessor
+import torch.nn.functional as F
+import torch
+from typing import List
+from tqdm import tqdm
+import nltk
+import string
+import numpy as np
+
```
Comment (maxime-louis) [inline models/context_processors/recomp.py:11]: 2 empty lines between functions/classes

File: models/context_processors/recomp.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        scores = {qi:{ci:{} for ci in range(len(contexts[qi]))} for qi in range(len(queries))}
+        for i, (qi, ci, si) in enumerate(zip(qis, cis, sis)):
+            scores[qi][ci][si] = (embs_q[qi] @ embs[i]).item()
+        selected_contexts = []
+        for qi in range(len(queries)):
+            selected_contexts_ = []
+            for ci in range(len(contexts[qi])):
+                scores_ = scores[qi][ci]
+                if self.threshold is None: # default
+                    selected_idxs = np.argsort(np.array([scores_[t+1] for t in range(len(scores_))]))[-self.top_k:]
```
Comment (maxime-louis) [inline models/context_processors/recomp.py:88]: t+1 indexing error ?

**Comment by maxime-louis** [comment]
> > just be to usre: here Provence does not do reranking right?
> 
> it does. If `reorder=True` in provence config, provence reorders documents in the context processing list. But it is incorporated into context processing, so you would set `retriever=<smth> context_processor=provence/provence_reranking_0.1` and skip reranker

Can you put an assert for that ?

File: models/context_processors/dslr_ce.py (hunk 2 above)
Comment (nadiinchi) [reply models/context_processors/dslr_ce.py:13]: thanks, good suggestion, added paper links in all methods!

File: models/context_processors/llmlingua2.py (hunk 6 above)
Comment (nadiinchi) [reply models/context_processors/llmlingua2.py:7]: done

File: models/context_processors/recomp.py (hunk 8 above)
Comment (nadiinchi) [reply models/context_processors/recomp.py:11]: done

File: models/context_processors/recomp.py (hunk 9 above)
Comment (nadiinchi) [reply models/context_processors/recomp.py:88]: same as is dslr, `scores_` is a dict which keys are shifted +1

**Comment by sclincha** [comment]
1) Do we want to update the requirements file for LLMLingua? Or Simply Try Catch, please install?
2)  It could be good to have a simple README for context processor explaining the list of models and basic usage such as:
`CONFIG=myconfig python bergen.py retriever=splade-v3  reranker=debertav3 +context_processor=provence/provence_standalone_0.1  dataset=$1 generator=vllm_llama-2-7b-chat`
 What do you think? Where should we put that README?. The run_exp file is a bit hard to read for a first usage ...

**Comment by nadiinchi** [comment]
> 1. Do we want to update the requirements file for LLMLingua? Or Simply Try Catch, please install?

@thibault-formal Can I ask you to deal with this? :)

> 2. It could be good to have a simple README for context processor explaining the list of models and basic usage such as:
>    `CONFIG=myconfig python bergen.py retriever=splade-v3  reranker=debertav3 +context_processor=provence/provence_standalone_0.1  dataset=$1 generator=vllm_llama-2-7b-chat`
>    What do you think? Where should we put that README?. The run_exp file is a bit hard to read for a first usage ...

I added a main Provence readme here: https://github.com/naver/bergen/tree/main/scripts/provence (also referenced in the paper)

**Comment by thibault-formal** [comment]
> > 1. Do we want to update the requirements file for LLMLingua? Or Simply Try Catch, please install?
> 
> @thibault-formal Can I ask you to deal with this? :)

I think @sclincha was initially worried it would lead to import errors for people not using llmlingua; but the import only happens when using this `LLMLingua2`/`LongLLMLingua` classes, so no real need to do anything, esp. bc now we have a proper Provence installation guide.
